qiskit/synthesis/discrete_basis/gate_sequence.py

```diff
@@ -0,0 +1,386 @@
+from __future__ import annotations
+
+from collections.abc import Sequence
+import math
+import numpy as np
+
+from qiskit.circuit import Gate, QuantumCircuit
+
+
+class GateSequence:
+    """A class implementing a sequence of gates.
+
+    This class stores the sequence of gates along with the unitary they implement.
+    """
+
+    def __init__(self, gates: Sequence[Gate] = ()) -> None:
+        """Create a new sequence of gates.
+
+        Args:
+            gates: The gates in the sequence. The default is [].
+        """
+        self.gates = list(gates)
+        self.matrices = [np.array(gate, dtype=np.complex128) for gate in gates]
+        self.labels = [gate.name for gate in gates]
+
+        # get U(2) representation of the gate sequence
+        u2_matrix = np.identity(2)
+        for matrix in self.matrices:
+            u2_matrix = matrix.dot(u2_matrix)
+
+        # convert to SU(2)
+        su2_matrix, global_phase = _convert_u2_to_su2(u2_matrix)
+
+        # convert to SO(3), that's what the Solovay Kitaev algorithm uses
+        so3_matrix = _convert_su2_to_so3(su2_matrix)
+
+        # store the matrix and the global phase
+        self._eulers = None
+        self.name = " ".join(self.labels)
+        self.global_phase = global_phase
+        self.product = so3_matrix
+        self.product_su2 = su2_matrix
+
+    def remove_cancelling_pair(self, indices: Sequence[int]) -> None:
+        """Remove a pair of indices that cancel each other and *do not* change the matrices."""
+        for index in list(indices[::-1]):
+            self.gates.pop(index)
+            self.labels.pop(index)
+
+        # restore name
+        self.name = " ".join(self.labels)
+
+    def __eq__(self, other: "GateSequence") -> bool:
+        """Check if this GateSequence is the same as the other GateSequence.
+
+        Args:
+            other: The GateSequence that will be compared to ``self``.
+
+        Returns:
+            True if ``other`` is equivalent to ``self``, false otherwise.
+
+        """
+        if not len(self.gates) == len(other.gates):
+            return False
+
+        for gate1, gate2 in zip(self.gates, other.gates):
+            if gate1 != gate2:
+                return False
+
+        if self.global_phase != other.global_phase:
+            return False
+
+        return True
+
+    def to_circuit(self):
+        """Convert to a circuit.
+
+        If no gates set but the product is not the identity, returns a circuit with a
+        unitary operation to implement the matrix.
+        """
+        if len(self.gates) == 0 and not np.allclose(self.product, np.identity(3)):
+            circuit = QuantumCircuit(1, global_phase=self.global_phase)
+            su2 = _convert_so3_to_su2(self.product)
+            circuit.unitary(su2, [0])
+            return circuit
+
+        circuit = QuantumCircuit(1, global_phase=self.global_phase)
+        for gate in self.gates:
+            circuit.append(gate, [0])
+
+        return circuit
```

Comment:
I wonder if it would make sense to have a `to_dag()` method here too so for the transpiler pass we can avoid `circuit_to_dag`. Thinking along the lines of: https://github.com/Qiskit/qiskit-terra/pull/9188

Reply:
Aha! I thought we should do all the evaluations on the DAG -- what you suggest above I can of course add 🙂